File: src/main/java/com/github/kayjamlang/core/provider/ValueExpressionProvider.java
package com.github.kayjamlang.core.provider;

import com.github.kayjamlang.core.Type;
import com.github.kayjamlang.core.exceptions.TypeException;
import com.github.kayjamlang.core.expressions.ValueExpression;

public class ValueExpressionProvider<A, B, C> extends ExpressionProvider<ValueExpression,
        A, B, C> {

    @Override
    public final Type getType(MainExpressionProvider<A, B, C> mainProvider,
                        B context,
                        B argsContext,
                        ValueExpression expression) throws TypeException {
        if(expression.value == null) {
            Type any = Type.NULL.clone();
            any.nullable = true;

            return any;
        }

        if(expression.value instanceof Double)
            return Type.DOUBLE;
        else if(expression.value instanceof Integer)
            return Type.INTEGER;
        else if(expression.value instanceof Long)
            return Type.LONG;
        else if(expression.value instanceof String)
            return Type.STRING;
        else if(expression.value instanceof Boolean)
            return Type.BOOLEAN;

        return Type.ANY;
    }
}

File: src/main/scala/com/github/kayjamlang/core/provider/ValueExpressionProvider.scala
package com.github.kayjamlang.core.provider

import com.github.kayjamlang.core.Type
import com.github.kayjamlang.core.exceptions.TypeException
import com.github.kayjamlang.core.expressions.ValueExpression
import com.github.kayjamlang.core.opcodes.AccessType

class ValueExpressionProvider[A, B, C](accessType: AccessType) extends ExpressionProvider[ValueExpression, A, B, C](accessType) {
  @throws[TypeException]
  override def getType(mainProvider: MainExpressionProvider[A, B, C], context: B, argsContext: B, expression: ValueExpression): Type = {
    val value = expression.value
    if (value == null) {
      val any = Type.NULL clone()
      any.nullable = true
      return any
    }

    value match {
      case _: Double  => Type DOUBLE
      case _: Int     => Type INTEGER
      case _: Long    => Type LONG
      case _: String  => Type STRING
      case _: Boolean => Type BOOLEAN
      case _          => Type ANY
    }
  }
}
